### PR TITLE
Add HMRC API support for the validation of GB VAT numbers

### PR DESCRIPTION
**Pull request by jarthod**
This pull request is one of the suggesting resulting from my work to integrate the HMRC API (see #99 for the discussion on the other options).

On this PR I tried to add support for HMRC API without changing too much existing code and the public API. I namespaced the VIES lookup code, added the HMRC version, tried to map the details output and exceptions to be close to the existing VIES format (so people can still use the same code on their side for GB and EU numbers). I added specs of course (including one for VIES too about network timeout because I wasn't sure of the behavior in this case). I also prepared the version bump, changelog update and readme update to save you time if we decide to merge this in the end.

The diff is a bit hard to read because of the namespacing causing file name change + some indentation (because of the Module) so they are considered removed and added files but actually have little changes.

### CODE REVIEW

File: lib/valvat/lookup.rb [hunk 1]
```diff
@@ -28,15 +31,18 @@ def valid?
     end
 
     def response
-      @response ||= Request.new(@vat, @options).perform
+      @response ||= case @vat.vat_country_code
+      when 'GB' then HMRC::Request.new(@vat, @options).perform
+      else VIES::Request.new(@vat, @options).perform
+      end
```
Comment (jarthod) [inline lib/valvat/lookup.rb:37]: This is where the main switching is done, for a "pluggable" version this code would need to be a bit more advanced but for now this is enough.

File: lib/valvat/lookup/vies/fault.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+          'VAT_BLOCKED' => BlockedError,
+          'IP_BLOCKED' => BlockedError,
+          'GLOBAL_MAX_CONCURRENT_REQ' => RateLimitError,
+          'GLOBAL_MAX_CONCURRENT_REQ_TIME' => RateLimitError,
+          'MS_MAX_CONCURRENT_REQ' => RateLimitError,
+          'MS_MAX_CONCURRENT_REQ_TIME' => RateLimitError
+        }.freeze
+
+        def self.fault_to_error(fault)
+          (FAULTS[fault] || UnknownLookupError).new("The VIES web service returned the error '#{fault}'.")
```
Comment (jarthod) [inline lib/valvat/lookup/vies/fault.rb:28]: Only real change in this file besides the namespacing is this line, I moved the exception custom error message here to make the custom exception classes more generic so they can be re-used by every service.

File: spec/valvat/lookup_spec.rb [hunk 3]
```diff
@@ -240,4 +304,132 @@
...
+        expect(result).to be(false)
+      end
+    end
+
+    describe 'Error : MESSAGE_THROTTLED_OUT' do
+      before {
+        stub_request(:get, /test-api.service.hmrc.gov.uk/).to_return(
+          status: 429,
+          body: '{"code":"MESSAGE_THROTTLED_OUT"}'
+        )
```
Comment (jarthod) [inline spec/valvat/lookup_spec.rb:335]: I had to add `webmock` to the dev dependencies to be able to test error responses as their test API do not offer some special VAT number to generate errors, they only offer different fake but valid numbers.

File: spec/valvat/lookup_spec.rb [hunk 4]
```diff
@@ -4,25 +4,54 @@
...
 describe Valvat::Lookup do
   describe '#validate' do
-    context 'with existing VAT number' do
+    context 'with existing EU VAT number' do
       it 'returns true' do
         result = described_class.validate('IE6388047V')
-        # Skip if VIES is down
-        expect(result).to be(true) unless result.nil?
+        skip "VIES is down" if result.nil?
+        expect(result).to be(true)
```
Comment (jarthod) [inline spec/valvat/lookup_spec.rb:11]: I changes a couple of these conditional to use the standard `skip` method from rspec this way when a test a skipped for this reason (service down) it actually shows in the test output, so the test is valid but you can see that it was skipped.

**Comment by pastullo** [comment]
Hey @jarthod thanks a lot for this PR.
I have been using it in production for quite some time. Would love to see this being merged into the master branch! ❤️

**Comment by pastullo** [comment]
@jarthod could you update this PR to v1.1.2 to fix the SOAPA error? There is already a merged PR to fix this issue.

File: spec/valvat/lookup_spec.rb (hunk 4 above)
Comment (jarthod) [reply spec/valvat/lookup_spec.rb:11]: @yolk I see you applied this in your `nethttp` branch so I'll mark this as resolved

File: spec/valvat/lookup_spec.rb (hunk 3 above)
Comment (jarthod) [reply spec/valvat/lookup_spec.rb:376]: Ok sorry again I found them in https://github.com/yolk/valvat/blob/9266fd143fea6eefb96ece9cde1f2efedcbbc887/spec/valvat/lookup_spec.rb :)
resolving this too :)

Comment (yolk) [reply spec/valvat/lookup_spec.rb:376]: No intention per se ... yepp, seems to be a good idea to test against these stubbed error cases. Thx!

Comment (yolk) [reply spec/valvat/lookup_spec.rb:376]: Parallel again :-) ... okay, I'll need to practice myself in some patience.

Comment (jarthod) [reply spec/valvat/lookup_spec.rb:376]: Mee too xD

**Comment by yolk** [comment]
Implemented in v1.2

**Comment by jarthod** [comment]
Thanks @yolk.
For people using my branch here (@pastullo?) I recommend to go back to the default gem with the `uk: true` option.
I'll remove this branch soon.

**Pull request closed without merging**